test/nfs-testdriver.go

```diff
@@ -0,0 +1,154 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/testsuites"
+)
+
+type nfsDriver struct {
+	driverInfo testsuites.DriverInfo
+	manifests  []string
+}
+
+var NFSdriver = InitNFSDriver
+
+type nfsVolume struct {
+	serverIP  string
+	serverPod *v1.Pod
+	f         *framework.Framework
+}
+
+// initNFSDriver returns nfsDriver that implements TestDriver interface
+func initNFSDriver(name string, manifests ...string) testsuites.TestDriver {
+	return &nfsDriver{
+		driverInfo: testsuites.DriverInfo{
+			Name:        name,
+			MaxFileSize: testpatterns.FileSizeLarge,
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+			Capabilities: map[testsuites.Capability]bool{
+				testsuites.CapPersistence: true,
+				testsuites.CapExec:        true,
+			},
+		},
+		manifests: manifests,
+	}
+}
+
+func InitNFSDriver() testsuites.TestDriver {
+
+	return initNFSDriver("csi-nfsplugin",
+		"csi-attacher-nfsplugin.yaml",
+		"csi-attacher-rbac.yaml",
+		"csi-nodeplugin-nfsplugin.yaml",
+		"csi-nodeplugin-rbac.yaml")
+
+}
+
+var _ testsuites.TestDriver = &nfsDriver{}
+var _ testsuites.PreprovisionedVolumeTestDriver = &nfsDriver{}
+var _ testsuites.PreprovisionedPVTestDriver = &nfsDriver{}
+
+func (n *nfsDriver) GetDriverInfo() *testsuites.DriverInfo {
+	return &n.driverInfo
+}
+
+func (n *nfsDriver) SkipUnsupportedTest(pattern testpatterns.TestPattern) {
+	if pattern.VolType == testpatterns.DynamicPV {
+		framework.Skipf("NFS Driver does not support dynamic provisioning -- skipping")
+	}
+}
+
+func (n *nfsDriver) GetPersistentVolumeSource(readOnly bool, fsType string, volume testsuites.TestVolume) (*v1.PersistentVolumeSource, *v1.VolumeNodeAffinity) {
+	nv, _ := volume.(*nfsVolume)
+	return &v1.PersistentVolumeSource{
+		CSI: &v1.CSIPersistentVolumeSource{
+			Driver:       n.driverInfo.Name,
+			VolumeHandle: "nfs-vol",
+			VolumeAttributes: map[string]string{
+				"server":   nv.serverIP,
+				"share":    "/",
+				"readOnly": "true",
+			},
+		},
+	}, nil
+}
+
+func (n *nfsDriver) PrepareTest(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
+	config := &testsuites.PerTestConfig{
+		Driver:    n,
+		Prefix:    "nfs",
+		Framework: f,
+	}
+
+	//Install the nfs driver from the manifests
+	cleanup, err := config.Framework.CreateFromManifests(nil, n.manifests...)
```

Comment:
Instead of creating a new nfs driver for every test case, can we bring up the nfs driver once in the BeforeSuite()?

Reply:
Currently the testdriver automatically creates a new namespace with a randomly generated name for each testcase, and I think the driver needs to be in that same namespace for it to be accessible by the tests. This would be hard to do in the BeforeSuite() if we don't know what the randomly generated namespace is called.

Reply:
I think if you don't use CreateFromManifests, you can just bring up the driver directly in some default namespace.  This is how we configure pd csi driver to run e2e externally and not bring up the driver for every test:

https://github.com/kubernetes/kubernetes/blob/51db0fe21a816b0818df571ad70386864d83c722/test/e2e/storage/drivers/csi.go#L444

Reply:
Ya that makes sense. So we would bring up the driver in a default namespace, instead of creating a new namespace for each test case right?

Reply:
Yeah the driver could be brought up in the default namespace.  Test case application pods will still come up in their own namespaces though

Reply:
Is there something wrong with bringing up the Test case application pods in the same default namespace as the driver? I'm trying to run them in separate namespaces (driver in default, and test case application pods in its own), but the test cases either timeout or take a long time to finish.

Reply:
In a real production cluster, the driver is generally installed in its own namespace and user applications will be in their own namespaces too.  So cross namespace access must be possible.  If you are seeing issues I would try to dig in more. It could indicate that the test is dependent on the driver deployment more than it needs to

Reply:
@msau42 I'm still testing it, but it seems to be working fine. Would it be better to have the user manually deploy the driver in a default namespace, instead of having the tests use these specific [YAML files](https://github.com/kubernetes-csi/csi-driver-nfs/pull/10/commits/892c7a866e88708d11f0d0902d3c45787c101277#diff-c69d8975ffcf100ed7be43c0052d993bR59) to do it? This way I think it would be easier for users to run e2e tests with their own private images.

Reply:
there are 2 situations, 1 is CI which is when one would want to redeploy `latest` yamls before every test, 2 is testing an already-installed driver. It would be nice if this example could demonstrate both? Is there a way for the test to detect if the driver is already installed?

Reply:
Even in a CI situation, would you want to deploy the yamls every time?  Nothing about the driver yamls should have changed in between each run?

Reply:
for now auto-deploying whatever is in the yaml folder every BeforeSuite is fine. If the user wants to test a different image version they can edit the yamls. In the future maybe there will be a flag for whether  the test should deploy the driver or not if it already exists

Reply:
@msau42 it's possible, like e.g. hostpath driver had some hardcoded paths that were changed so the yaml changed

Reply:
oh, I misspoke didn't mean before every test i mean before every testsuite. We agree on that already

Reply:
It looks like this is still bringing up the nfs driver for every test case?

Reply:
Sorry, I still haven't pushed it.